great_expectations/core/expectation_configuration.py

```diff
@@ -1273,6 +1264,35 @@ def get_evaluation_parameter_dependencies(self):
         dependencies = _deduplicate_evaluation_parameter_dependencies(dependencies)
         return dependencies
 
+    def _traverse_expectation_configuration_kwargs(
+        self, root: Any, parsed_dependencies: dict = {}
+    ) -> dict:
+        """
+        Helper method that treats ExpectationConfiguration kwargs as a tree and uses it
+        to enable a recursive, DFS traversal.
+
+        Designed to account for $PARAMETERS at any level of the kwargs dict.
+        """
+        # We want to step through standard iterables
+        if isinstance(root, (tuple, list)):
+            for child in root:
+                self._traverse_expectation_configuration_kwargs(
+                    child, parsed_dependencies
+                )
+        # dicts are similar but we need to account for $PARAMETERs
+        elif isinstance(root, dict):
+            for key, value in root.items():
+                if "$PARAMETER" in key:
+                    param_string_dependencies = find_evaluation_parameter_dependencies(
+                        value
+                    )
+                    nested_update(parsed_dependencies, param_string_dependencies)
+                self._traverse_expectation_configuration_kwargs(
+                    value, parsed_dependencies
+                )
+        # Any other types is discarded, ending a branch of the traversal and unwinding the call stack
+        return parsed_dependencies
+
```

Comment:
The recursive solution is definitely more elegant that the iterative one here but I could see both being reasonably difficult to understand. Unfortunately, traversal of a nested object like the kwargs here is a problem that requires tree/graph traversal (an inherently complicated topic).

I've tried to comment the overall flow but if there's any confusion here, please let me know and we can work together to simplify things. I opted for recursion because it's more readable and our kwargs should only be a few layers deep (thus removing any concern about large call stacks).